### PR TITLE
| OAI AKOA | Perceiver Transformer | 46252665 |

### PR DESCRIPTION
**Pull request by roycek7**
Hi @shakes76 and tutors,

This pull request implements the Perceiver Transformer on the Oai Akoa Knee Dataset using Tensorflow and Keras backend. The folder contains all the necessary scripts along with **__init__.py** (folder and file paths) files, **.yml** (environment file), **.sh** (training on goliath cluster), **.gitignore**, **perceiver_slurm_output.txt** (best result obtained) and **requirements.txt** (dependencies). The **main.py** driver file is responsible for the processing of the dataset, training the perceiver transformer, and displaying the patches. The perceiver model is situated in the perceiver folder separated by its components. The training class compiles the perceiver which sequentially constructs the rest of its components. The script and the folder structure are as follows:

### Folder Structure
* data
    *  Akoa_Analysis (Images)
    *  processed (.npy files)
* display
    * figures
* logs (save model weights)
* perceiver 
* pre_process (dataset processor)
* settings (hyper-parameters)

### Scripts
-  main.py: main driver file to pre-process, train, and plot patches
- display/display_plot.py: plots random sample from dataset and draws patches
- pre_process/dataset_processor.py:     
    * Creates Labels
    * Augments Data (switch-on function call for processing dataset with augmentation)
    * Splits into Train, Test
    * Saves as .npy files
- settings/config.py: the hyper-parameters and configurations for augmentation, labels, and models, etc
- perceiver/attention_module.py: cross attention module and transformer block
- perceiver/extract_patches.py: extracts patches from the input array
- perceiver/feed_forward.py: sequential neural network with dense layers
- perceiver/load_data.py: loads the pre-processed .npy data files
- perceiver/patch_encoder.py: creates positional encoding for input data
- perceiver/perceiver.py: creates the architecture of perceiver with cross attention and transformers block followed by global average pooling and feed-forward network for classifications heads.
- perceiver/train.py: trains the perceiver on the dataset and plots the loss/accuracy

Please see README.md for a complete description of the architecture, results, and plots, including the perceivers dependencies and hyper-parameters.

Any feedback is welcome!

Cheers,
Pritish Kumar Roy
46252665

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
Difficulty Hard

Functionality: OK


Code
- Structure:  OK

README 
- Content: OK


PR Target OK

General Comments
- Commit messages: OK
- Excellent

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present
No Data leakage 
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No feedback required
Request Description very good
